web/src/compose_fade.js

```diff
@@ -68,7 +68,7 @@ function fade_messages() {
 
     // Update the visible messages first, before the compose box opens
     for (i = 0; i < visible_groups.length; i += 1) {
-        $first_row = rows.first_message_in_group(visible_groups[i]);
+        $first_row = rows.first_message_in_group($(visible_groups[i]));
```

Comment:
The `$` here is a function being called, not a “prefix” like your commit message suggests.

Reply:
Fixed it. Thanks for pointing out !.

Reply:
Hmm I suspect something is a bit wrong with the types in `rows.ts`; does `message_viewport.visible_groups` return a jQuery object in the first place? I think it might, but it's a bit hard to tell, because there's some weird stuff in that code path, like the `row_to_id` function that definitely returns a `Message` object, not a row:
```
    function row_to_id(row: HTMLElement): Message {                                                     
        assert(message_lists.current !== undefined);                                                    
        return message_lists.current.get(rows.id($(row)))!;                                             
    }                                                                                                   
```

Can you investigate that a bit and maybe add some prep commits making it more sane?

Reply:
I think we have some confusion here. 

`message_viewport.visible_groups` has a local function `get_row` which gets passed as argument to fn `_visible_divs `which subsequently passes it to fn `add_to_visible` .So `add_to_visible` has this function as parameter with name `row_to_id` and calls it inside the function body.

```

       if (in_viewport_or_tall(row_rect, top_of_feed, bottom_of_feed, require_fully_visible)) {
            visible.push(row_to_id(row));
        } 

```


For the code path I described this `row_to_id` actually refers to the` get_row` function(as mentioned ,inside the `visible_groups`  **which does return `row`**) not the `row_to_id` fn  (inside `visible_messages` function which return `Message` object)  that you sampled.

The confusion mainly arises due to naming of the `row_to_id` parameter( in `add_to_visible` fn signature)  which is part of two different code paths. In one it refers to `get_row` (inside `visible_groups`)and in other it refers to `row_to_id`(inside `visible_messages` fn).

Reply:
I have added a prep commit renaming callback `row_to_id` to `row_to_output`. This is bit more self-explanatory as the callback returns a generic type(not always an `id` or a `row`) and is also consistent with name `_visible_divs` uses for it.